File: dashboard/src/stockbit.ts
/**
 * Stockbit API client for the web dashboard.
 * Uses the Vite dev server proxy at /api/stockbit to bypass CORS
 * and inject the STOCKBIT_JWT_TOKEN securely.
 */

const BASE_URL = '/api/stockbit';

export async function fetchMarketDetector(emiten: string, fromDate: string, toDate: string): Promise<any> {
  const url = new URL(`${BASE_URL}/marketdetectors/${emiten}`, window.location.origin);
  url.searchParams.append('from', fromDate);
  url.searchParams.append('to', toDate);
  url.searchParams.append('transaction_type', 'TRANSACTION_TYPE_NET');
  url.searchParams.append('market_board', 'MARKET_BOARD_REGULER');
  url.searchParams.append('investor_type', 'INVESTOR_TYPE_ALL');
  url.searchParams.append('limit', '25');

  const response = await fetch(url.toString(), {
    method: 'GET',
    headers: { 'accept': 'application/json' },
  });

  if (!response.ok) throw new Error(`Stockbit Market Detector API error: ${response.status}`);
  return response.json();
}

export async function fetchOrderbook(emiten: string): Promise<any> {
  const url = `${BASE_URL}/company-price-feed/v2/orderbook/companies/${emiten}`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Orderbook API error: ${response.status}`);
  return response.json();
}

export async function fetchEmitenInfo(emiten: string): Promise<any> {
  const url = `${BASE_URL}/emitten/${emiten}/info`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Emiten Info API error: ${response.status}`);
  return response.json();
}

export async function fetchSectors(): Promise<string[]> {
  const url = `${BASE_URL}/emitten/sectors`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Sectors API error: ${response.status}`);
  const data = await response.json();
  return (data.data || []).map((item: { name: string }) => item.name).filter(Boolean);
}

export async function fetchWatchlistGroups(): Promise<any[]> {
  const url = `${BASE_URL}/watchlist?page=1&limit=500`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Watchlist Groups API error: ${response.status}`);
  const json = await response.json();
  return Array.isArray(json.data) ? json.data : [];
}

export async function fetchWatchlist(watchlistId?: number): Promise<any> {
  let id = watchlistId;
  if (!id) {
    const groups = await fetchWatchlistGroups();
    const defaultGroup = groups.find((w: any) => w.is_default) || groups[0];
    id = defaultGroup?.watchlist_id;
    if (!id) throw new Error('No watchlist found');
  }

  const detailUrl = `${BASE_URL}/watchlist/${id}?page=1&limit=500`;
  const response = await fetch(detailUrl, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Watchlist Detail API error: ${response.status}`);
  
  const json = await response.json();
  if (json.data?.result) {
    json.data.result = json.data.result.map((item: any) => ({
      ...item,
      company_code: item.symbol || item.company_code
    }));
  }
  return json;
}

export async function fetchKeyStats(emiten: string): Promise<any> {
  const url = `${BASE_URL}/keystats/ratio/v1/${emiten}?year_limit=10`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`KeyStats API error: ${response.status}`);
  
  const json = await response.json();
  const categories = json.data?.closure_fin_items_results || [];
  
  const findCategory = (name: string) => {
    const category = categories.find((c: any) => c.keystats_name === name);
    if (!category) return [];
    return category.fin_name_results.map((r: any) => r.fitem);
  };

  return {
    currentValuation: findCategory('Current Valuation'),
    incomeStatement: findCategory('Income Statement'),
    balanceSheet: findCategory('Balance Sheet'),
    profitability: findCategory('Profitability'),
    growth: findCategory('Growth'),
  };
}

export async function fetchHistoricalSummary(
  emiten: string,
  startDate: string,
  endDate: string,
  limit: number = 12
): Promise<any[]> {
  const url = `${BASE_URL}/company-price-feed/historical/summary/${emiten}?period=HS_PERIOD_DAILY&start_date=${startDate}&end_date=${endDate}&limit=${limit}&page=1`;
  const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Historical Summary API error: ${response.status}`);
  const json = await response.json();
  return json.data?.result || [];
}

export async function deleteWatchlistItem(watchlistId: number, companyId: number): Promise<void> {
  const url = `${BASE_URL}/watchlist/${watchlistId}/company/${companyId}/item`;
  const response = await fetch(url, { method: 'DELETE', headers: { 'accept': 'application/json' } });
  if (!response.ok) throw new Error(`Delete Watchlist Item API error: ${response.status}`);
}

export function getTopBroker(marketDetectorData: any): any | null {
  const brokers = marketDetectorData?.data?.broker_summary?.brokers_buy;
  if (!brokers || !Array.isArray(brokers) || brokers.length === 0) return null;
  const topBroker = [...brokers].sort((a, b) => Number(b.bval) - Number(a.bval))[0];
  return {
    bandar: topBroker.netbs_broker_code,
    barangBandar: Math.round(Number(topBroker.blot)),
    rataRataBandar: Math.round(Number(topBroker.netbs_buy_avg_price)),
  };
}

export function getBrokerSummary(marketDetectorData: any): any {
  const detector = marketDetectorData?.data?.bandar_detector;
  const brokerSummary = marketDetectorData?.data?.broker_summary;
  return {
    detector: {
      top1: detector?.top1 || { vol: 0, percent: 0, amount: 0, accdist: '-' },
      top3: detector?.top3 || { vol: 0, percent: 0, amount: 0, accdist: '-' },
      top5: detector?.top5 || { vol: 0, percent: 0, amount: 0, accdist: '-' },
      avg: detector?.avg || { vol: 0, percent: 0, amount: 0, accdist: '-' },
      total_buyer: detector?.total_buyer || 0,
      total_seller: detector?.total_seller || 0,
      number_broker_buysell: detector?.number_broker_buysell || 0,
      broker_accdist: detector?.broker_accdist || '-',
      volume: detector?.volume || 0,
      value: detector?.value || 0,
      average: detector?.average || 0,
    },
    topBuyers: brokerSummary?.brokers_buy?.slice(0, 4) || [],
    topSellers: brokerSummary?.brokers_sell?.slice(0, 4) || [],
  };
}

/** Fetch active trade data — recent daily trading activity for a ticker */
export async function fetchActiveTrade(emiten: string, days: number = 10): Promise<any[]> {
  const end = new Date();
  const start = new Date();
  start.setDate(start.getDate() - (days + 5)); // extra buffer for weekends
  const fmt = (d: Date) => d.toISOString().split('T')[0];
  return fetchHistoricalSummary(emiten, fmt(start), fmt(end), days);
}

/** Fetch Stockbit community stream / news for a ticker */
export async function fetchStockbitNews(emiten: string): Promise<any[]> {
  try {
    const url = `${BASE_URL}/stream/stocks/${emiten}?page=1&limit=15`;
    const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
    if (!response.ok) return [];
    const json = await response.json();
    const posts = json.data?.stream || json.data || [];
    return Array.isArray(posts) ? posts.slice(0, 15) : [];
  } catch {
    return [];
  }
}

/** Fetch IDX news from Stockbit's news endpoint */
export async function fetchIdxNews(): Promise<any[]> {
  try {
    const url = `${BASE_URL}/news?page=1&limit=20`;
    const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
    if (!response.ok) return [];
    const json = await response.json();
    return Array.isArray(json.data) ? json.data.slice(0, 20) : [];
  } catch {
    return [];
  }
}

/** Fetch daily broker flow / bandarmology from tradersaham API */
export async function fetchBrokerFlowDaily(emiten: string, days: number = 60): Promise<any> {
  try {
    const url = `/api/tradersaham/api/market-insight/broker-intelligence?limit=100&page=1&sort_by=consistency&mode=accum&lookback_days=${days}&broker_status=Bandar,Whale,Retail,Mix&search=${emiten.toLowerCase()}`;
    const response = await fetch(url, { method: 'GET', headers: { 'accept': 'application/json' } });
    if (!response.ok) return null;
    return await response.json();
  } catch {
    return null;
  }
}

File: stockbit.py
import os
import logging
import requests
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

STOCKBIT_BASE_URL = 'https://exodus.stockbit.com'
STOCKBIT_AUTH_URL = 'https://stockbit.com'

class StockbitClient:
    """Client for extracting data from Stockbit API."""

    def __init__(self, token=None):
        self.token = token or os.getenv("STOCKBIT_JWT_TOKEN")
        if not self.token:
            logger.warning("STOCKBIT_JWT_TOKEN is not set. API calls will likely fail.")

    def _get_headers(self) -> dict:
        return {
            'accept': 'application/json',
            'authorization': f'Bearer {self.token}',
            'origin': 'https://stockbit.com',
            'referer': 'https://stockbit.com/',
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36',
            'accept-language': 'en-GB,en-US;q=0.9,en;q=0.8,id;q=0.7',
            'sec-ch-ua': '"Not(A:Brand";v="8", "Chromium";v="144", "Google Chrome";v="144"',
            'sec-ch-ua-mobile': '?0',
            'sec-ch-ua-platform': '"Windows"',
        }

    def _handle_response(self, response: requests.Response, api_name: str) -> dict:
        if response.status_code == 401:
            logger.error(f"{api_name}: Token expired or invalid (401)")
            # You could optionally raise a custom TokenExpiredError here
            response.raise_for_status()
        
        try:
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            logger.error(f"{api_name} HTTP error: {e}")
            raise
        except ValueError as e:
            logger.error(f"{api_name} JSON decode error: {e}")
            raise

    def fetch_market_detector(self, ticker: str, from_date: str, to_date: str) -> dict | None:
        """Fetch market detector (bandarmology) data for a given ticker and date range."""
        if not self.token:
            return None

        url = f"{STOCKBIT_BASE_URL}/marketdetectors/{ticker}"
        params = {
            'from': from_date,
            'to': to_date,
            'transaction_type': 'TRANSACTION_TYPE_NET',
            'market_board': 'MARKET_BOARD_REGULER',
            'investor_type': 'INVESTOR_TYPE_ALL',
            'limit': '25'
        }

        try:
            response = requests.get(url, headers=self._get_headers(), params=params, timeout=10)
            return self._handle_response(response, 'Market Detector API')
        except Exception as e:
            logger.error(f"Failed to fetch market detector for {ticker}: {e}")
            return None

    def fetch_orderbook(self, emiten: str) -> dict | None:
        """Fetch Orderbook data (market data)"""
        url = f"{STOCKBIT_BASE_URL}/company-price-feed/v2/orderbook/companies/{emiten}"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            return self._handle_response(response, 'Orderbook API')
        except Exception as e:
            logger.error(f"Failed to fetch orderbook for {emiten}: {e}")
            return None

    def fetch_emiten_info(self, emiten: str) -> dict | None:
        """Fetch Emiten Info (including sector)"""
        url = f"{STOCKBIT_BASE_URL}/emitten/{emiten}/info"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            return self._handle_response(response, 'Emiten Info API')
        except Exception as e:
            logger.error(f"Failed to fetch emiten info for {emiten}: {e}")
            return None

    def fetch_sectors(self) -> list[str]:
        """Fetch all sectors list"""
        url = f"{STOCKBIT_BASE_URL}/emitten/sectors"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            data = self._handle_response(response, 'Sectors API')
            sectors = [item.get('name') for item in data.get('data', []) if item.get('name')]
            return sectors
        except Exception as e:
            logger.error(f"Failed to fetch sectors: {e}")
            return []

    def fetch_watchlist_groups(self) -> list[dict]:
        """Fetch all watchlist groups"""
        url = f"{STOCKBIT_BASE_URL}/watchlist?page=1&limit=500"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            data = self._handle_response(response, 'Watchlist Groups API')
            return data.get('data', []) if isinstance(data.get('data'), list) else []
        except Exception as e:
            logger.error(f"Failed to fetch watchlist groups: {e}")
            return []

    def fetch_watchlist(self, watchlist_id: int = None) -> dict | None:
        """Fetch Watchlist data by ID (or default if not provided)"""
        try:
            if not watchlist_id:
                groups = self.fetch_watchlist_groups()
                default_group = next((w for w in groups if w.get('is_default')), groups[0] if groups else None)
                if not default_group:
                    raise ValueError('No watchlist found')
                watchlist_id = default_group.get('watchlist_id')

            url = f"{STOCKBIT_BASE_URL}/watchlist/{watchlist_id}?page=1&limit=500"
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            json_data = self._handle_response(response, 'Watchlist Detail API')

            # Map symbol to company_code for compatibility
            if json_data.get('data', {}).get('result'):
                for item in json_data['data']['result']:
                    item['company_code'] = item.get('symbol') or item.get('company_code')

            return json_data
        except Exception as e:
            logger.error(f"Failed to fetch watchlist {watchlist_id}: {e}")
            return None

    def fetch_key_stats(self, emiten: str) -> dict | None:
        """Fetch KeyStats data for a stock"""
        url = f"{STOCKBIT_BASE_URL}/keystats/ratio/v1/{emiten}?year_limit=10"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            json_data = self._handle_response(response, 'KeyStats API')
            
            categories = json_data.get('data', {}).get('closure_fin_items_results', [])
            
            def find_category(name: str):
                category = next((c for c in categories if c.get('keystats_name') == name), None)
                if not category:
                    return []
                return [r.get('fitem') for r in category.get('fin_name_results', [])]

            return {
                'currentValuation': find_category('Current Valuation'),
                'incomeStatement': find_category('Income Statement'),
                'balanceSheet': find_category('Balance Sheet'),
                'profitability': find_category('Profitability'),
                'growth': find_category('Growth'),
            }
        except Exception as e:
            logger.error(f"Failed to fetch keystats for {emiten}: {e}")
            return None

    def fetch_historical_summary(self, emiten: str, start_date: str, end_date: str, limit: int = 12) -> list[dict]:
        """Fetch historical price summary from Stockbit"""
        url = f"{STOCKBIT_BASE_URL}/company-price-feed/historical/summary/{emiten}?period=HS_PERIOD_DAILY&start_date={start_date}&end_date={end_date}&limit={limit}&page=1"
        try:
            response = requests.get(url, headers=self._get_headers(), timeout=10)
            json_data = self._handle_response(response, 'Historical Summary API')
            return json_data.get('data', {}).get('result', [])
        except Exception as e:
            logger.error(f"Failed to fetch historical summary for {emiten}: {e}")
            return []

    def delete_watchlist_item(self, watchlist_id: int, company_id: int) -> bool:
        """Delete item from watchlist"""
        url = f"{STOCKBIT_BASE_URL}/watchlist/{watchlist_id}/company/{company_id}/item"
        try:
            response = requests.delete(url, headers=self._get_headers(), timeout=10)
            self._handle_response(response, 'Delete Watchlist Item API')
            return True
        except Exception as e:
            logger.error(f"Failed to delete watchlist item {company_id} from {watchlist_id}: {e}")
            return False

def get_top_broker(market_detector_data: dict) -> dict | None:
    """Extract the top broker by net buy value."""
    if not market_detector_data:
        return None
        
    try:
        data = market_detector_data.get('data', {})
        broker_summary = data.get('broker_summary', {})
        brokers_buy = broker_summary.get('brokers_buy', [])
        
        if not brokers_buy:
            return None
            
        sorted_brokers = sorted(brokers_buy, key=lambda x: float(x.get('bval', 0)), reverse=True)
        top_broker = sorted_brokers[0]
        
        return {
            'bandar': top_broker.get('netbs_broker_code'),
            'barangBandar': round(float(top_broker.get('blot', 0))),
            'rataRataBandar': round(float(top_broker.get('netbs_buy_avg_price', 0)))
        }
    except Exception as e:
        logger.error(f"Error parsing top broker: {e}")
        return None

def get_broker_summary(market_detector_data: dict) -> dict | None:
    """Extract the bandar detector summary."""
    if not market_detector_data:
        return None
        
    try:
        data = market_detector_data.get('data', {})
        detector = data.get('bandar_detector', {})
        broker_summary = data.get('broker_summary', {})
        
        return {
            'detector': {
                'top1': detector.get('top1', {'vol': 0, 'percent': 0, 'amount': 0, 'accdist': '-'}),
                'top3': detector.get('top3', {'vol': 0, 'percent': 0, 'amount': 0, 'accdist': '-'}),
                'top5': detector.get('top5', {'vol': 0, 'percent': 0, 'amount': 0, 'accdist': '-'}),
                'avg': detector.get('avg', {'vol': 0, 'percent': 0, 'amount': 0, 'accdist': '-'}),
                'total_buyer': detector.get('total_buyer', 0),
                'total_seller': detector.get('total_seller', 0),
                'number_broker_buysell': detector.get('number_broker_buysell', 0),
                'broker_accdist': detector.get('broker_accdist', '-'),
                'volume': detector.get('volume', 0),
                'value': detector.get('value', 0),
                'average': detector.get('average', 0),
            },
            'topBuyers': broker_summary.get('brokers_buy', [])[:4],
            'topSellers': broker_summary.get('brokers_sell', [])[:4],
        }
    except Exception as e:
        logger.error(f"Error parsing broker summary: {e}")
        return None
